From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pancakse <pancakseyt@gmail.com>
Date: Fri, 19 Jan 2024 22:40:55 -0700
Subject: [PATCH] Optimize CraftServer.getWorld(UUID)

Original code by MultiPaper - https://github.com/MultiPaper/MultiPaper which is licensed under the MIT

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 9dfee38e6e65e34b11f1875f23c5c330977bb670..156fb6067e9ecd42062fb03c81586449b2320ec4 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -263,6 +263,8 @@ import net.md_5.bungee.api.chat.BaseComponent; // Spigot
 import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
 
+import it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap; // IceCream
+
 public final class CraftServer implements Server {
     private final String serverName = "IceCream"; // Paper // Pufferfish // Purpur // IceCream
     private final String serverVersion;
@@ -279,6 +281,7 @@ public final class CraftServer implements Server {
     protected final DedicatedServer console;
     protected final DedicatedPlayerList playerList;
     private final Map<String, World> worlds = new LinkedHashMap<String, World>();
+    private final Map<UUID, World> worldsByUUID = new Object2ObjectLinkedOpenHashMap<>(); // IceCream - MultiPaper - optimize getWorld(UUID)
     private final Map<Class<?>, Registry<?>> registries = new HashMap<>();
     private YamlConfiguration configuration;
     private YamlConfiguration commandsConfiguration;
@@ -1462,6 +1465,7 @@ public final class CraftServer implements Server {
             this.getLogger().log(Level.SEVERE, null, ex);
         }
 
+        this.worldsByUUID.remove(world.getUID()); // IceCream - MultiPaper - optimize getWorld(UUID)
         this.worlds.remove(world.getName().toLowerCase(java.util.Locale.ENGLISH));
         this.console.removeLevel(handle);
         return true;
@@ -1480,6 +1484,7 @@ public final class CraftServer implements Server {
 
     @Override
     public World getWorld(UUID uid) {
+        if (true) return this.worldsByUUID.get(uid); // IceCream - MultiPaper - optimize getWorld(UUID)
         for (World world : this.worlds.values()) {
             if (world.getUID().equals(uid)) {
                 return world;
@@ -1503,6 +1508,7 @@ public final class CraftServer implements Server {
             System.out.println("World " + world.getName() + " is a duplicate of another world and has been prevented from loading. Please delete the uid.dat file from " + world.getName() + "'s world directory if you want to be able to load the duplicate world.");
             return;
         }
+        this.worldsByUUID.put(world.getUID(), world); // IceCream - MultiPaper - optimize getWorld(UUID)
         this.worlds.put(world.getName().toLowerCase(java.util.Locale.ENGLISH), world);
     }
 
